src/ngraph/op/fused/shuffle_channels.cpp

```diff
@@ -0,0 +1,107 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "ngraph/op/fused/shuffle_channels.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/op/util/reshape.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+op::ShuffleChannels::ShuffleChannels(const shared_ptr<Node>& data,
+                                     const int axis,
+                                     const size_t groups)
+    : FusedOp("ShuffleChannels", {data})
+    , m_groups{groups}
+{
+    if (axis < 0)
+    {
+        m_axis = axis + data->get_shape().size();
+    }
+    else
+    {
+        m_axis = axis;
+    }
+
+    constructor_validate_and_infer_types();
+}
+
+void op::ShuffleChannels::pre_validate_and_infer_types()
+{
+    const auto shape = get_argument(0)->get_shape();
+
+    NODE_VALIDATION_CHECK(
+        this, shape.size() >= 1, "The input tensor's shape is expected to be at least 1D.");
+
+    NODE_VALIDATION_CHECK(this,
+                          m_axis >= 0 && m_axis < shape.size(),
+                          "The 'axis' parameter for ShuffleChannels has to point to one of the "
+                          "input tensor's shape dimensions.");
+
+    const auto channel_dim_size = shape.at(m_axis);
+    NODE_VALIDATION_CHECK(
+        this,
+        channel_dim_size % m_groups == 0,
+        "The channel dimension size has to be a multiple of the groups parameter value.");
+}
+
+NodeVector op::ShuffleChannels::decompose_op() const
+{
+    const auto data = get_argument(0);
+    const auto& data_shape = data->get_shape();
+
+    const auto reshaped = util::reshape(data, get_pre_shuffle_shape(data_shape));
+    const auto shuffled = util::reorder_axes(reshaped, {0, 2, 1, 3});
```

Comment:
will we need to add shape.size() == 4 to pre_validate_and_infer_types?

Reply:
I don't think so. I've tested it locally and it works for any tensor that is 1D or higher. An intermediate step for this operation is reshaping to 4D and shuffling data in this format. Then it's reshaped back to the original dimensions.